### ISSUE TITLE
Critical error while generating reStructured text files

### ISSUE THREAD

**Issue by goodlux** [svenevs/exhale#33]
I'm trying to exhale on a large code base that isn't particularly well documented. The doxygen XML gets created, but the .rst file generation is running into issues with file names that are too long

[+] Exhale: finished parsing Doxygen XML in 20.77 seconds.
[~] Exhale: generating reStructuredText documents.

OSError: [Errno 63] File name too long: (!) Critical error while generating the file for [docs/in/api/function_caffe2__operator_takes_as_input_a_tensor_of_floats_to_find_the_percentile_values_for_COMMA__a_2D_tensor_of_floats_COMMA__where_the_first_column_of_the_tensor_represents_sampled_values_COMMA__and_the_second_column_represents_the_percentile_labels_COMMA__and_a_tensor_of_integers_lengths._This_lengths_tensor_is_used_because_the_operator_works_on_multiple_sets_of_raw_values_at_the_same_time._For_example_COMMA__for_an_input_original_values=_COMMA__lengths=[2_COMMA__1_COMMA__1]_COMMA__value_to_pct=Our_operator_expects_that_each_column_i_of_the_input_tensor_is_sampled_from_distribution_i._Lengths_tells_us_that_the_first_two_elements_in_value_to_pct_are_sampled_from_distribution_1_COMMA__the_next_is_from_distribution_two_COMMA__and_the_last_is_from_distribution_3._We_expect_the_output_of_our_operator_to_give_us_._To_calculate_the_percentile_of_an_element_COMMA__we_check_to_see_if_its_value_is_already_mapped_to_a_percentile_in_value_to_pct._If_so_COMMA__we_return_that_value._If_not_COMMA__we_linearly_interpolate_between_the_two_closest_values_in_value_to_pct._If_the_value_is_larger_than_all_values_in_value_to_pct_COMMA__we_return_1._If_it_'s_smaller_than_all_the_values_COMMA__we_return_0._RPAREN__DOC"_RPAREN__.Input.rst]

I'm not sure where to fix the specfic documentation, or the Doxygen config, or the Sphinx config to allow it to handle this. Is there some way to force it to ignore this particular file and keep building?

**Comment by svenevs** [comment]
Like you may be able to find the documentation with selective grepping. Most of the underscores there are from spaces, so maybe `grep -ri "Our operator expects that" include/` or something. Just don't make the phrase too long because if the docstrings you search for is split across lives `grep` won't find it.

**Comment by goodlux** [comment]
I'm running it on this directory:
 https://github.com/pytorch/pytorch/tree/master/caffe2/operators 
I can't tell which file it is choking on though. Also, trying to use different Doxygen directives from 
"exhaleDoxygenStdin", but I'm not sure what format put these in.

**Comment by goodlux** [comment]
Also, I'm running this on a mac, so the long filename might be an OS dependent issue?

**Comment by svenevs** [comment]
Technically this is operating system dependent, but every filesystem I know of has a maximum file length encoded, typically 1024 or 2048 characters.

1. The problem comes from here: https://github.com/pytorch/pytorch/blob/1ca6e77615ac0020df97ca237b1af542ad1d4bf7/caffe2/operators/percentile_op.cc#L85-L106
2. The circumstances of your code base may not be amenable to Exhale.  Doxygen will process all of these files, so both `.h` files and `.cc` files.  There are a few things at play here:

    - Doxygen has trouble with macros that are not called with a semicolon.  So even simple macros get skipped, and you'll (very very unfortunately) need to predefine all of the metaprogramming you are doing.
        - Example of how to do that with `exhaleDoxygenStdin` [on the companion site](https://github.com/svenevs/exhale-companion/blob/13335ccf60bf4f703aefafd5eb5d8bfcdaf3a7b4/docs/conf.py#L64-L65).
    - What makes this hard is I don't really know which version of [`OPERATOR_SCHEMA`](https://github.com/pytorch/pytorch/blob/1ca6e77615ac0020df97ca237b1af542ad1d4bf7/caffe2/core/operator_schema.h#L509) is going to be used.  And even if you did, enumerating all of the expansions for every item would probably be an exercise in futility.

To be honest, I don't think Doxygen can handle your framework, meaning Exhale / Breathe cannot either.  You can add a line `GENERATE_HTML = YES` to `exhaleDoxygenStdin`.  Next to the `xml/` directory there should be an `html` directory, you should open that up and see what doxygen is doing, but most likely those docs are also completely broken.

Some possible workarounds:

1. At the expense of having no documentation, skip anything that is breaking.  Exhale inserts two compile-time directives to help with this:

    https://github.com/svenevs/exhale/blob/d5a6adbc9481a02c49ad5540bc30b7bfd62d886a/exhale/configs.py#L898-L899

    There's no anchor for it, but scroll to the `PREDEFINED` section at the bottom [of the "mastering" doxygen docs](http://exhale.readthedocs.io/en/latest/mastering_doxygen.html#additional-variables-with-important-impacts).  The idea is to `#ifdef` away code that breaks docs compilation (e.g., forward declarations).

2. That won't really save you, because in the end the metaprogramming here with raw string literal docs is great for python, but not for C++.  Specifically, Doxygen will never pick those up as documentation strings, since docstrings have to be comments starting before the declaration (`/**`, `/!`, or `///`).  Changing the metaprogramming for adding documentation is technically possible, but would be a huge pain in the ass and probably break all of the python documentation (since even if you got it working, you won't be able to get that into raw string literals now for the python docs).

    So basically the main friction here is that the current documentation is setup for python help strings, and as a consequence no C++ documentation parser will succeed in extracting those.  Traditionally, you have all of the docstrings in `.h` files.  So you could add to `exhaleDoxygenStdin` something like `EXCLUDE_PATTERNS = *.cc *.cu` to make it only process the `.h` files.

    But then you'd need to add documentation to all of the `.h` files.

I hope that wasn't too much information.  The main problem you have here is the tension between "documentability" in C++ vs Python help strings.  Solving the C++ problem, unfortunately, will result in a large amount of duplicated documentation, one for the C++ docstring, and one for Python.  I think that will ultimately be a recipe for disaster.

Although I hate to discourage you from using Exhale,

1. Metaprogramming and Doxygen is hard to get right, and the way I create filenames creates problems for SFINAE stuff.  I need to use proper C++ name mangling for the filenames, or else SFINAE templates [will also generate names that are too long](https://github.com/svenevs/exhale/pull/12#issuecomment-362873856) :/
2. Where very large projects are concerned, Breathe will actually be a bottleneck for you (especially on ReadTheDocs).  There's an extensive discussion about that on [`breathe/#315`](https://github.com/michaeljones/breathe/issues/315), but the gist: the breathe parser needs to be updated because of memory leaks with `minidom`.  I've actually started work on updating it, but it's a beast of an update that I won't be able to fix for quite some time.

All that said, I'm happy to help you hack through things if you want to keep going with it!  Feel free to ask more questions / clarifications on the above!

**Comment by goodlux** [comment]
Definitely not too much info, all of this is very helpful.

**Comment by goodlux** [comment]
Hi @svenevs! Circling back to this; I wanted to reach out to you about some related questions, but this probably isn't the best thread. How can I contact you directly? My email: rob.kunkle@gmail.com

Thanks!

**Comment by ddemidov** [comment]
Could the issue be solved by changing how the file names are generated? I got the same error when generating docs for a heavily templated library. Changing node file names from mangled name to SHA1 of the mangled names seems to solve the issue for me:
```diff
diff --git a/exhale/graph.py b/exhale/graph.py
index 938679b..7d823f8 100644
--- a/exhale/graph.py
+++ b/exhale/graph.py
@@ -18,6 +18,7 @@ import sys
 import codecs
 import itertools
 import textwrap
+import hashlib
 from bs4 import BeautifulSoup
 
 try:
@@ -2021,7 +2022,10 @@ class ExhaleRoot(object):
         # create the file and link names
         node.file_name = os.path.join(
             self.root_directory,
-            "{kind}_{name}.rst".format(kind=node.kind, name=html_safe_name)
+            "{sha}.rst".format(sha=hashlib.sha1(
+                "{kind}_{name}".format(
+                    kind=node.kind, name=html_safe_name).encode()
+                ).hexdigest())
         )
         node.link_name = "{kind}_{name}".format(
             kind=utils.qualifyKind(node.kind).lower(),
```

**Comment by svenevs** [comment]
Hi @ddemidov,

Thanks for resurfacing this, I need to fix this ASAP. Is your project public? I want to keep human friendly filenames where possible, but for simplicity templates will get special cased because the current scheme is absurdly naive.  Doxygen already produces a unique hash, so I want to use that.

I'd like to test using your code if possible to make sure the intended fix actually works before doing a release.

**Comment by ddemidov** [comment]
The project in question is https://github.com/ddemidov/amgcl; I've just setup a branch that tests exhale here: https://github.com/ddemidov/amgcl/tree/docs-exhale, https://github.com/ddemidov/amgcl/commit/1d399ab6a5319cfdc05bd974210da10914361c89.

**Comment by svenevs** [comment]
**Excellent**, thank you very much! I should be able to solve this today over and for all. It keeps falling off my radar and shouldn't because the fixes won't be hard.

**Comment by svenevs** [comment]
@ddemidov see #35 for the proposed fix, the reason it is not as simple as the sha is for reasons you would never have known about (what I do / do not use from Breathe, and what is actually linkable via sphinx).

I wanted to mention that while I was able to get past the error for you on your project, there are some other things going on with your code.

1. What you told `exhale` in `conf.py` about `"doxygenStripFromPath"` is not what you told Doxygen (since you are running it on your own via the `Doxyfile`).  This results in things like directories having the absolute path, which is probably not what you want (especially on RTD...).

    I suggest you take a look at whether or not Exhale running Doxygen for you would work out.  It seems like you are testing out potential doc system changes, using `"exhaleDoxygenStdin"` is the encouraged approach.  It also already adds the `\rst` and `\endrst` to `ALIASES` for you automatically :wink:

2. I must will point that unfortunately the state of affairs with Exhale and templates (because of Breathe) is really frustrating.  You will get over 1000 warnings about the same thing.  It's "safe" to ignore in that things will get documented at show up somewhere, but they may have weird `template <> template <...>` things going on because of Breathe.  I am working with Breathe to fix this, but TBH I may just abandon it and solve it differently.

    So in reality, this means that actual documentation warnings that you would ordinarily want to squash will be harder to find.  I wish this were not the case, but it is what it is at least for now :cry:

3. Function overloads.  I want to solve that too, as I just need to parse parameters.  Your framework is large / I couldn't tell if this affects you, but I will fix that before the `0.2.0` release (which has the filename too long fixes as well).  Just wanted to note that this may come up and it is a known problem to look out for (#13)

Basically, Exhale was able to create _a_ tree view for your project, but the file hierarchy was wrong (mis-coordination between Doxygen and Exhale), and it is unclear to me whether or not the treeView was actually correct.  While I am hopeful that Exhale will work for your project, it may be too "impolished" at this point in time.  Please feel free to raise additional issues with concrete examples of things Exhale messes up with your framework.  I cannot guarantee how quickly I will be able to solve things depending on the problem, but maybe in the future Exhale might be more polished / ready for heavily templated code like yours.

**Comment by svenevs** [comment]
Cool!  My absurd test cases don't work on Windows because of their antiquated filesystem.  I'm gonna have to introduce some added trickery to fix that, but I'm hoping to finish it before I sleep tonight...

**Comment by svenevs** [comment]
@goodlux FYI this issue being closed is a win / lose for you.  Win in the sense that PyTorch definitely needs the improved filename logic.  Lose in the sense that the original issue readily identified that something was going seriously wrong with what Exhale thought the name of things was (the whole using the documentation as the name of the documented object thing) **will no longer be readily obvious** as of current `master` and imminent `0.2.0` release.

Just pointing it out so that you know there may now be "hidden bugs" because I fallback on the `sha1` approach suggested here: https://github.com/svenevs/exhale/issues/33#issuecomment-407364460
